Data/HashMap/Base.hs

```diff
@@ -431,22 +442,59 @@ member k m = case lookup k m of
 -- | /O(log n)/ Return the value to which the specified key is mapped,
 -- or 'Nothing' if this map contains no mapping for the key.
 lookup :: (Eq k, Hashable k) => k -> HashMap k v -> Maybe v
-lookup k0 m0 = go h0 k0 0 m0
-  where
-    h0 = hash k0
-    go !_ !_ !_ Empty = Nothing
-    go h k _ (Leaf hx (L kx x))
-        | h == hx && k == kx = Just x  -- TODO: Split test in two
-        | otherwise          = Nothing
-    go h k s (BitmapIndexed b v)
-        | b .&. m == 0 = Nothing
-        | otherwise    = go h k (s+bitsPerSubkey) (A.index v (sparseIndex b m))
-      where m = mask h s
-    go h k s (Full v) = go h k (s+bitsPerSubkey) (A.index v (index h s))
-    go h k _ (Collision hx v)
-        | h == hx   = lookupInArray k v
-        | otherwise = Nothing
-{-# INLINABLE lookup #-}
+lookup k m = lookup' (hash k) k 0 m
+{-# INLINE lookup #-}
+
+-- Internal helper for lookup. This version takes the precomputed hash so
+-- that functions that make multiple calls to lookup and related functions
+-- (insert, delete) only need to calculate the hash once.
+lookup' :: (Eq k, Hashable k) => Hash -> k -> Int -> HashMap k v -> Maybe v
+lookup' !_ !_ !_ Empty = Nothing
+lookup' h k _ (Leaf hx (L kx x))
+    | h == hx && k == kx = Just x  -- TODO: Split test in two
+    | otherwise          = Nothing
+lookup' h k s (BitmapIndexed b v)
+    | b .&. m == 0 = Nothing
+    | otherwise    = lookup' h k (s+bitsPerSubkey) (A.index v (sparseIndex b m))
+  where m = mask h s
+lookup' h k s (Full v) = lookup' h k (s+bitsPerSubkey) (A.index v (index h s))
+lookup' h k _ (Collision hx v)
+    | h == hx   = lookupInArray k v
+    | otherwise = Nothing
+{-# INLINABLE lookup' #-}
+
+-- Internal helper for lookup. This version takes the precomputed hash so
+-- that functions that make multiple calls to lookup and related functions
+-- (insert, delete) only need to calculate the hash once.
+--
+-- It is used by 'alterF' so that hash computation and key comparison only needs
+-- to be performed once. With this information you can use the more optimized
+-- versions of insert ('insertNewKey', 'insertExistingKeyNoCollision',
+-- 'insertExistingKeyCollision') and delete ('deleteNoCollision',
+-- 'deleteCollision').
+--
+-- Outcomes:
+--   Key not in map           => Nothing
+--   Key in map, no collision => Just (v, Nothing)
+--   Key in map, collision    => Just (v, position)
+lookupRecordCollision
+  :: (Eq k, Hashable k)
+  => Hash -> k -> Int -> HashMap k v -> Maybe (v, Maybe Int)
```

Comment:
Done.

Another option would be to combine the `Alone` and `Collided` case and just set `-1` here. So you'd have:
```haskell
data LookupRes a = Absent | Present a !Int
```

Where `noCollision v = Present v (-1)` and `withCollision v collPos = Present v collPos`. This would simplify the `alterF` implementation by not having to case match on `Alone` and `Collided` since the alone case we need to pass `-1` anyways. Instead you would have in the case of delete:

```haskell
case lookupRes of
  Absent -> m
  Present _ collPos = deleteKeyExists collPos h k 0 m
```

What do you think?

Reply:
Sorry, I guess that should be

```haskell
data LookupRes a = Absent | Alone a | Collided !Int a
```

Or even just

```haskell
data LookupRes a = Absent | Present !Int a
```

with negative values meaning no collision. That's probably better.